gaffer-as-a-service/gaas-rest/src/main/java/uk/gov/gchq/gaffer/gaas/factories/ProxyStoreType.java

```diff
@@ -0,0 +1,68 @@
+/*
+ * Copyright 2020 Crown Copyright
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.gchq.gaffer.gaas.factories;
+
+import org.springframework.stereotype.Service;
+import uk.gov.gchq.gaffer.common.model.v1.GafferSpec;
+import uk.gov.gchq.gaffer.gaas.model.GaaSCreateRequestBody;
+import uk.gov.gchq.gaffer.proxystore.ProxyStore;
+import java.util.HashMap;
+import java.util.Map;
+import static uk.gov.gchq.gaffer.gaas.util.Constants.STORE_PROPERTIES_KEY;
+import static uk.gov.gchq.gaffer.proxystore.ProxyProperties.GAFFER_CONTEXT_ROOT;
+import static uk.gov.gchq.gaffer.proxystore.ProxyProperties.GAFFER_HOST;
+import static uk.gov.gchq.gaffer.store.StoreProperties.STORE_CLASS;
+
+@Service
+public class ProxyStoreType implements StoreType {
+
+    @Override
+    public String getType() {
+        return "proxyStore";
+    }
+
+    @Override
+    public AbstractStoreTypeBuilder getStoreSpecBuilder(final GaaSCreateRequestBody graph) {
+        return new ProxyStoreSpecBuilder(graph);
+    }
+
+    private static class ProxyStoreSpecBuilder extends AbstractStoreTypeBuilder {
+
+        ProxyStoreSpecBuilder(final GaaSCreateRequestBody graph) {
+            this.graph = graph;
+        }
+
+        private Map<String, Object> getDefaultProxyStoreProperties(final String host, final String contextRoot) {
+            final Map<String, Object> proxyStoreProperties = new HashMap<>();
+            proxyStoreProperties.put(STORE_CLASS, ProxyStore.class.getName());
+            proxyStoreProperties.put(GAFFER_HOST, host);
+            if (contextRoot != null) {
+                proxyStoreProperties.put(GAFFER_CONTEXT_ROOT, contextRoot);
+                // else, let Gaffer handle the default context root when not specified in GaaS REST request
+            }
+            return proxyStoreProperties;
+        }
+
+        @Override
+        public GafferSpec build() {
+            final GafferSpec gafferSpec = super.build();
+            gafferSpec.putNestedObject(getDefaultProxyStoreProperties(graph.getProxyHost(), graph.getProxyContextRoot()), STORE_PROPERTIES_KEY);
```

Comment:
Rather than having all the individual variables such as `proxyHost` and `proxyContextRoot` you could have a `storeProperties` `Map` containing all of these properties. You could add the method `setStoreProperties` to the `AbstractStoreTypeBuilder` and at `build` time you could search for the properties expected and if they don't exist throw an error such as `StoreTypeException`.

The `storeProperties` `Map` could be passed to the `rest` service `GaaSCreateRequestBody` taking the place of `proxyHost` and `proxyContextRoot`.  These values should be stored as key/values in the `storeProperties` `Map` instead. If you need to add values for new `StoreTypes` in the future then you could add them to the `storeProperties` rather than add new variables such as `proxyHost` and `proxyContextRoot`.

Reply:
@rocky341, if you think this change is too much work then I can create a new ticket for it instead?

Reply:
@r32575 might be best to make this is a separate ticket if that's okay.

Reply:
@r32575 Actually will add to this PR

Reply:
This is now done